Relational: Translate SelectMany

smitpatel opened the pull request:
SelectMany in linq look something like following (second clause is considered the collection selector)
```C#
from c in cs
from o in c.Orders
select....
```
Following are the translations of SelectMany

Unrelated collection selector
```C#
from c in cs
from o in os
```
Generates CROSS JOIN

Correlated collection selector with correlation being a predicate
```C#
from c in cs
from o in os.Where(o => o.CustomerID == c.CustomerID)
```
Such predicate can be lifted and used in generating a join. So this query generates JOIN.
If collection selector ends with DefaultIfEmpty then it is LEFT JOIN.

Correlated collection selector with correlation not a predicate
```C#
from c in cs
from o in os.Select(o => c.City)
```
Since we cannot generate a join predicate here, this translates to CROSS APPLY.
If collection selector ends with DefaultIfEmpty then it is OUTER APPLY.

- Add support for Cross Apply
- Add Support for Outer Apply
- Convert Cross Apply to Inner Join when possible
- Convert Outer Apply to Left Join when possible
- Add translation for DefaultIfEmpty
- Add translation for both overloads of SelectMany
- Handle DefaultIfEmpty & SelectMany without collectionSelector in Navigation Expansion
- Ensure columns are in projection when generating join predicate from a correlated subquery

Currently there are no tests for Cross/Outer Apply.
Our earlier Cross Apply got converted to join. N+1 evaluation tests are disabled right now, which would generate Cross Apply.
We never supported Outer Apply in past.

Resolves #15711
Resolves #12567
Resolves #12572
Resolves #12872
Resolves #16330
Resolves #15081
Resolves #16989

Re-enable tests for #12449

ajcvickers commented:
@maumar Please make it a priority to review this when you get in. (You were nominated in triage! 🚎)

maumar commented on test/EFCore.SqlServer.FunctionalTests/Query/OwnedQuerySqlServerTest.cs:
```diff
@@ -225,16 +225,17 @@ public override void Query_when_subquery()
 LEFT JOIN [OwnedPerson] AS [o6] ON [t0].[Id] = [o6].[Id]
 LEFT JOIN [OwnedPerson] AS [o7] ON [o6].[Id] = [o7].[Id]
-ORDER BY [t0].[Id]");
+LEFT JOIN [Order] AS [o8] ON [t0].[Id] = [o8].[ClientId]
```
why extra join here?

smitpatel replied:
Orders was owned collection and it was ignored earlier.

maumar commented on test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs:
```diff
@@ -6505,7 +6493,7 @@ public override async Task Cast_to_derived_type_after_OfType_works(bool isAsync)
                 @"SELECT [g].[Nickname], [g].[SquadId], [g].[AssignedCityName], [g].[CityOrBirthName], [g].[Discriminator], [g].[FullName], [g].[HasSoulPatch], [g].[LeaderNickname], [g].[LeaderSquadId], [g].[Rank]
 FROM [Gears] AS [g]
-WHERE [g].[Discriminator] = N'Officer'");
+WHERE [g].[Discriminator] IN (N'Gear', N'Officer') AND ([g].[Discriminator] = N'Officer')");
```
we should optimize this at some point (later)

smitpatel replied:
You mean #15808 :trollface:

maumar commented on src/EFCore.SqlServer/Query/Internal/SearchConditionConvertingExpressionVisitor.cs:
```diff
@@ -317,6 +317,26 @@ protected override Expression VisitCrossJoin(CrossJoinExpression crossJoinExpres
             return crossJoinExpression.Update(table);
         }
 
+        protected override Expression VisitCrossApply(CrossApplyExpression crossApplyExpression)
```
logic repeated 3 times - dry

smitpatel replied:
There is no good way to do it. Update function does not exist on abstract class.

smitpatel commented:
Renamed CrossApplyExpression -> InnerJoinLateralExpression & OuterApplyExpression -> LeftJoinLateralExpression to match with formal SQL terminology. Also updated relational implementation and moved SqlServer specific sql syntax in provider code.
